src/kernel/arch/aarch64/arch.zig

```diff
@@ -120,5 +109,133 @@ pub fn haltNoInterrupts() noreturn {
+        \\ mov x2, #0x08000000
+        \\ orr x1, x1, x2
+        \\ str w1, [x0, #0x08]
+        \\ mov x1, #0x20000000
+        \\ str w1, [x0, #0x28]
+        :
+        :
+        : "x0", "x1", "x2"
+    );
+}
+
+pub fn spinLed(period: u32) noreturn {
+    const activity_led = 29;
+    pinSetFunction(activity_led, .Output);
+    var i: u32 = 0;
+    var j: u32 = 0;
+    while (true) : (j += 1) {
+        if (j % (period * 1000 / 2) == 0) {
+            i += 1;
+            pinWrite(activity_led, @truncate(u1, i));
+        }
+    }
+}
+
+fn pinSetFunction(pin_index: u6, f: PinFunction) void {
+    mmio.readModifyWriteField(mmio_addr, .GPIO_FSEL0, @enumToInt(f), pin_index);
+}
+
+fn pinSetPull(pin_index: u6, pull: PinPull) void {
+    mmio.write(mmio_addr, .GPIO_PUD, @enumToInt(pull));
+    delay(150);
+    mmio.writeClock(mmio_addr, .GPIO_PUDCLK0, @as(u1, 1), pin_index);
+    delay(150);
+    mmio.writeClock(mmio_addr, .GPIO_PUDCLK0, @as(u1, 0), pin_index);
+    mmio.write(mmio_addr, .GPIO_PUD, @enumToInt(PinPull.None));
+}
+
+fn pinWrite(pin_index: u6, zero_or_one: u1) void {
+    mmio.writeClock(mmio_addr, if (zero_or_one == 0) mmio.Register.GPIO_CLR0 else .GPIO_SET0, @as(u1, 1), pin_index);
+}
+
+const PinFunction = enum(u3) {
+    Input = 0,
+    Output = 1,
+    AlternateFunction0 = 4,
+    AlternateFunction1 = 5,
+    AlternateFunction2 = 6,
+    AlternateFunction3 = 7,
+    AlternateFunction4 = 3,
+    AlternateFunction5 = 2,
+};
+
+const PinPull = enum {
+    None,
+    Down,
+    Up,
+};
+
+fn delay(count: usize) void {
+    var i: usize = 0;
+    while (i < count) : (i += 1) {
+        asm volatile ("mov x0, x0");
+    }
+}
+
+// map 1GB to ram except last 16MB to mmio
+pub fn enableFlatMmu() void {
+    const level_2_table = @intToPtr([*]usize, 0x50000)[0..2];
+    level_2_table[0] = 0x60000 | 0x3;
+    level_2_table[1] = 0x70000 | 0x3;
+    const page_table = @intToPtr([*]usize, 0x60000)[0 .. 2 * 8 * 1024];
+    const page_size = 64 * 1024;
+    const start_of_mmio = page_table.len - (16 * 1024 * 1024 / page_size);
+    var index: usize = 0;
+    while (index < start_of_mmio) : (index += 1) {
+        page_table[index] = index * page_size + 0x0703; // normal pte=3 attr index=0 inner shareable=3 af=1
+    }
+    while (index < page_table.len) : (index += 1) {
+        page_table[index] = index * page_size + 0x0607; // device pte=3 attr index=1 outer shareable=2 af=1
+    }
+
+    asm volatile (
+        \\ mov x3, #0x04ff
+        \\ msr mair_el3, x3
+        \\
+        \\ mrs x3, tcr_el3
+        \\ mov x2, #0x4022 // tcr_el3 t0sz 34 tg0 64k
+        \\ orr x3, x3, x2
+        \\ mov x2, #0x80800000
+        \\ orr x3, x3, x2
+        \\ msr tcr_el3, x3
+        \\
+        \\ mov x1, #0x50000
+        \\ msr ttbr0_el3, x1
+        \\ isb
+        \\ mrs x0, sctlr_el3
+        \\ orr x0, x0, #1
+        \\ msr sctlr_el3, x0
+        \\ isb
+        :
+        :
+        : "x0", "x1", "x2", "x3"
+    );
```

Comment:
Same again regarding the raw assembly code.

Reply:
This is no longer assembly but the constants still need explanation.